docs(block-node): add Block Node Operator FAQ (#3158)
hiero-ledger/hiero-block-node: docs/block-node/faq/operator-faq.md

> # Block Node Operator FAQ

Reviewer: probably good to add a link in `docs/block-node/README.md`

| **5005**            | JVM remote debug (JDWP)                              | Inbound from debugger                           | **Must be denied in production** — enabling JDWP significantly degrades performance |
| Outbound **40840**  | Backfill — BN dials peer BN's Subscribe API          | Outbound to peer BNs                            | Required if backfill plugin is enabled                                              |

The Block Node does not terminate TLS in-process; TLS is handled upstream by a Kubernetes
Ingress or load balancer.

> See [Network Ports and Protocols](./operations/network-ports-and-protocols.md) for full details.

### What are the expected inbound and outbound traffic flows?

**Inbound (port 40840):**
- Consensus Nodes push block streams (Tier 1 only — requires `stream-publisher` plugin)
- Mirror Nodes and downstream Block Nodes subscribe to block streams
- All clients call `serverStatus` to discover available block ranges
- Kubernetes kubelet issues HTTP GET liveness and readiness probes

**Inbound (port 16007):**
- Prometheus / monitoring scrapes the OpenMetrics endpoint

> **Outbound (port 40840):**

Reviewer 2: Outbound ports are not fixed; the outbound port is dynamically assigned when a connection is opened.

- Backfill plugin dials peer Block Nodes to fetch missing historical blocks

> The Block Node does not initiate any other outbound connections.

Reviewer 2: The Block Node _does_ connect to mirror node for RSA address book information when running the RSA Bootstrap Plugin. This is needed for RSA verification of wrapped record blocks.

| **Status (40840)**              | Not supported until a qualified CN release is available.                                                                                                                                                           |
| **Metrics (16007)**             | Internal only — do not expose publicly. TLS is not relevant.                                                                                                                                                       |

**Authentication:** There is no built-in authentication mechanism, and there are no
plans to add any. Security is enforced at the network and transport layer.

> See [Network Ports and Protocols](./operations/network-ports-and-protocols.md) and [Block Node On-Chain Registration](./block-node-on-chain-registration.md) for full details.

### How should I secure the Block Node if there is no built-in authentication?

The Block Node has no authentication and there are no plans to add any. This is by
design: **trust is in the data, not in the node.** Every block carries a
[Block Proof](./glossary.md#block-proof) that cryptographically verifies the block's

> authenticity — subscribers verify the data itself rather than trusting the node

Reviewer 2: ```suggestion
authenticity — subscribers verify the data themselves rather than trusting the node
```

### What is the difference between node types (Full Node, Rolling-History, Light Node, Archive Server)?

|        Type         |                     History retained                      |                          Typical use                          |
|---------------------|-----------------------------------------------------------|---------------------------------------------------------------|
| **Full Node**       | All history from genesis on local storage                 | Tier 1 mainnet — `plugin-profile-lfh` or `plugin-profile-all` |
| **Rolling-History** | Recent history only (configurable window, e.g. 7–90 days) | Tier 2 — low-cost redistribution                              |
| **Light Node**      | Minimal — health and status only                          | Development, testing, testnet                                 |
| **Archive Server**  | Cold storage — no live streaming                          | Offline archival                                              |

> See [Block Node Types and Tiers](../../Block-Node-Types.md) for full details.

### How do I deploy a Block Node?

Three paths are available:

> 1. **Solo Provisioner (recommended for testnet/evaluation):** Single command handles VM

Reviewer 2: My understanding is that Solo Provisioner is recommended for mainnet Tier 1.
Among other things, Solo Provisioner automates many of the necessary networking and traffic shaping tasks based on dynamic values provided by the managed Block Node.

   provisioning, Kubernetes setup, and Helm installation.
   See [Deploy with Solo Provisioner](./operations/solo-weaver-single-node-k8s-deployment.md).

2. **Bare-metal Kubernetes (recommended for mainnet Tier 1):** Manual Helm install on a
   pre-existing single-node cluster using the `task helm-release` Taskfile target.

>    See [Bare Metal Single Node Kubernetes Deployment](./operations/single-node-k8s-deployment.md).

Reviewer 2: I don't think this is recommended for mainnet tier 1, but it does remain an option.

3. **Existing Kubernetes cluster:** Apply the Block Node Helm chart directly with
   `-f charts/block-node-server/values-overrides/plugin-profile-lfh.yaml` (or your
   chosen profile) and any site-specific overrides.

### What plugin configuration do I need?

Select a pre-built Helm values override from
[`charts/block-node-server/values-overrides/`](https://github.com/hiero-ledger/hiero-block-node/tree/main/charts/block-node-server/values-overrides):

|         Profile          |                      Use                       |
|--------------------------|------------------------------------------------|
| `plugin-profile-lfh`     | Tier 1 — full history on local NVMe + HDD      |
| `plugin-profile-rfh`     | Remote archival — cloud storage backend        |

> | `plugin-profile-all`     | Full history local + cloud backup              |

Reviewer 2: This profile is only intended for testing, it includes plugins that may conflict with each other and produce unexpected results.

### What are the liveness and readiness probe URLs?

|   Probe   |             Default URL              | HTTP method |
|-----------|--------------------------------------|-------------|
| Liveness  | `http://<host>:40840/healthz/livez`  | GET         |
| Readiness | `http://<host>:40840/healthz/readyz` | GET         |

Both paths are configurable via `blockNode.health.liveness.endpoint` and
`blockNode.health.readiness.endpoint` in Helm values.

> See [Configuration Reference](./configuration.md) for full details.

### What metrics should I alert on?

**Medium severity — page on call:**

|                        Metric                        |   Threshold   |
|------------------------------------------------------|---------------|
| `blocknode_app_state_status`                         | ≠ 1 (RUNNING) |
| `blocknode_publisher_receive_latency_ns`             | > 20 seconds  |
| `blocknode_verification_blocks_error`                | > 3 in 60 s   |
| `blocknode_publisher_block_send_response_failed`     | > 3 in 60 s   |
| `blocknode_publisher_stream_errors`                  | > 3 in 60 s   |

> | `blocknode_files_recent_persistence_time_latency_ns` | > 20 seconds  |

Reviewer 2: A couple of these are a bit too high to be useful or low enough to be noisy.

```suggestion
| `blocknode_app_state_status`                         | ≠ 1 (RUNNING)      |
| `blocknode_publisher_receive_latency_ns`             | > 10 seconds       |
| `blocknode_verification_blocks_error`                | > 3 in 60 s        |
| `blocknode_publisher_block_send_response_failed`     | > 5 in 60 s        |
| `blocknode_publisher_stream_errors`                  | > 5 in 60 s        |
| `blocknode_files_recent_persistence_time_latency_ns` | > 20 milliseconds  |
```

> See [Configuration Reference](./configuration.md) for full details.

### How do I configure or change what Block Nodes are used as backfill sources?

Set `BACKFILL_BLOCK_NODE_SOURCES_PATH` to the path of a JSON file listing peer Block
Nodes:

```json
{
  "nodes": [
    {
      "address": "tier1-bn.example.com",
      "port": 40840,
      "priority": 0,
      "name": "Primary archive"
    }
  ]
}
```

All Tier 1 Block Nodes should have at least one source configured. More than one is
recommended for redundancy — the backfill plugin selects by earliest available block,
then priority, then health score.

> See [Preparing for WRB Cutover](./operations/preparing-your-block-node-for-wrb-cutover.md) for full details.

---

## Costs and economics

### Who pays for ingress and egress costs?

**Operators pay infrastructure costs directly** — their cloud provider credit card is on
file and billed for all ingress, egress, and compute.

**Hedera provides daily rewards** intended to offset operational expenses including
bandwidth and hardware costs. These rewards are not guaranteed to cover all costs —
if the reward amount does not cover actual spend, the operator absorbs the difference.

> **Co-location is strongly recommended.** Placing a Block Node in the same data centre

Reviewer 2: British spelling (not sure if we're standardizing on British or American):

```suggestion
**Co-location is strongly recommended.** Placing a Block Node in the same data center
```

| `state_service.proto`                  | `StateService.stateSnapshot` — *(defined; not yet implemented)*                                                 |
| `proof_service.proto`                  | `ProofService` — block content and state proofs *(not yet implemented)*                                         |
| `reconnect_service.proto`              | `ReconnectService.reconnect()` — provides state + block data to lagging Consensus Nodes *(not yet implemented)* |
| `network-data.proto`                   | Shared network endpoint message types: `NetworkData`, `NetworkConnection`                                       |
| `shared_message_types.proto`           | Shared message types: `BlockItemSet`, `BlockProof`, `EndOfStream`, etc.                                         |

The BN API protos are defined locally in `protobuf-sources/src/main/proto/block-node/api/`
within this repository. The Consensus Node protos (pulled for combined artifact generation)
originate from the
[hiero-ledger/hiero-consensus-node](https://github.com/hiero-ledger/hiero-consensus-node)

> repository and are fetched by `protobuf-sources/scripts/build-bn-proto.sh`.

Reviewer 2: Block Node also publishes a release artifact for every release with the full set of protocol buffer definitions (.proto files) supported by that release of the block node.

|       Plugin name        |                                       Feature provided                                       |             Tier required             |
|--------------------------|----------------------------------------------------------------------------------------------|---------------------------------------|
| `stream-publisher`       | Accepts block streams from Consensus Nodes (`publishBlockStream` RPC)                        | Tier 1 only — **remove for Tier 2**   |
| `stream-subscriber`      | Serves block streams to Mirror Nodes and downstream Block Nodes (`subscribeBlockStream` RPC) | Tier 1 and Tier 2                     |
| `block-access-service`   | Single-block random-access retrieval (`getBlock` RPC)                                        | Tier 1 and Tier 2                     |
| `server-status`          | `serverStatus` and `serverStatusDetail` RPCs — block range, version, plugin list             | All deployments                       |
| `health`                 | Kubernetes liveness (`/healthz/livez`) and readiness (`/healthz/readyz`) probes              | All deployments                       |

> | `verification`           | Verifies block proofs before persistence (TSS and RSA/WRB)                                   | All deployments                       |

Reviewer 2: The verification plugin name changed for `0.39` (new plugin version with significant rework, improvements, and fixes). This version also no longer accepts any block without cryptographic signature verification.
```suggestion
| `block-verification`     | Verifies block proofs before persistence (TSS and RSA/WRB)                                   | All deployments                       |
```

2. **Direct Single Node Kubernetes (an option for operators with an existing cluster):** Manual
   Helm install on a pre-existing single-node cluster using the `task helm-release` Taskfile target.
   See [Direct Single Node Kubernetes Deployment](./operations/single-node-k8s-deployment.md).

3. **Existing Kubernetes cluster:** Apply the Block Node Helm chart directly with
   `-f charts/block-node-server/values-overrides/plugin-profile-lfh.yaml` (or your
   chosen profile) and any site-specific overrides.

### What plugin configuration do I need?

Select a pre-built Helm values override from
[`charts/block-node-server/values-overrides/`](https://github.com/hiero-ledger/hiero-block-node/tree/main/charts/block-node-server/values-overrides):

|         Profile          |                      Use                       |
|--------------------------|------------------------------------------------|
| `plugin-profile-lfh`     | Tier 1 — full history on local NVMe + HDD      |
| `plugin-profile-rfh`     | Remote archival — cloud storage backend        |
| `plugin-profile-all`     | Full history local + cloud backup *(testing only — plugins may conflict and produce unexpected results)* |

> | `plugin-profile-minimal` | Development / testnet — health and status only |

Reviewer 2: Spotless is complaining about this table because it insists on padding every row...

### What are the liveness and readiness probe URLs?

|   Probe   |             Default URL              | HTTP method |
|-----------|--------------------------------------|-------------|
| Liveness  | `http://<host>:40840/healthz/livez`  | GET         |
| Readiness | `http://<host>:40840/healthz/readyz` | GET         |

Both paths are configurable via `blockNode.health.liveness.endpoint` and
`blockNode.health.readiness.endpoint` in Helm values.

> See [Configuration Reference](./configuration.md) for full details.

### What metrics should I alert on?

**Medium severity — page on call:**

|                        Metric                        |   Threshold   |
|------------------------------------------------------|---------------|
| `blocknode_app_state_status`                         | ≠ 1 (RUNNING) |
| `blocknode_publisher_receive_latency_ns`             | > 10 seconds      |
| `blocknode_verification_blocks_error`                | > 3 in 60 s       |
| `blocknode_publisher_block_send_response_failed`     | > 5 in 60 s       |
| `blocknode_publisher_stream_errors`                  | > 5 in 60 s       |

> | `blocknode_files_recent_persistence_time_latency_ns` | > 20 milliseconds |

Reviewer 2: Spotless is complaining here as well.

|        Type         |                     History retained                      |                          Typical use                          |
|---------------------|-----------------------------------------------------------|---------------------------------------------------------------|
| **Full Node**       | All history from genesis on local storage                 | Tier 1 mainnet — `plugin-profile-lfh` or `plugin-profile-all` |
| **Rolling-History** | Recent history only (configurable window, e.g. 7–90 days) | Tier 2 — low-cost redistribution                              |
| **Light Node**      | Minimal — health and status only                          | Development, testing, testnet                                 |
| **Archive Server**  | Cold storage — no live streaming                          | Offline archival                                              |

> See [Block Node Types and Tiers](../../Block-Node-Types.md) for full details.

### How do I deploy a Block Node?

Three paths are available:

1. **Solo Provisioner (recommended for mainnet Tier 1 and testnet/evaluation):** Single

>    command handles VM provisioning, Kubernetes setup, and Helm installation. Solo Provisioner
>    also automates networking and traffic shaping tasks based on dynamic values from the managed
>    Block Node.
>    See [Deploy with Solo Provisioner](./operations/solo-weaver-single-node-k8s-deployment.md).

Reviewer 2: Solo Provisioner works for both VM and Bare Metal, but does not provision a VM.

```suggestion
1. **Solo Provisioner (recommended for mainnet Tier 1 and testnet/evaluation):** Single
   command handles Kubernetes setup, and Helm installation. Solo Provisioner also automates
   networking and traffic shaping tasks based on dynamic values from the managed Block Node.
   See [Deploy with Solo Provisioner](./operations/solo-weaver-single-node-k8s-deployment.md).
```

2. **Bare-metal Kubernetes (an option for operators with an existing cluster):** Manual Helm
   install on a pre-existing single-node cluster using the `task helm-release` Taskfile target.

>    See [Bare Metal Single Node Kubernetes Deployment](./operations/single-node-k8s-deployment.md).

Reviewer 2: One item of feedback here, `Bare Metal` implies installation directly on hardware, which this is not; it's agnostic to whether the machines are hardware or VM.  We should probably rename to "Direct Single Node Kubernetes Deployment" rather than "Bare Metal Single Node Kubernetes Deployment"
